pkg/storegateway/indexheader/reader_pool.go

```diff
@@ -214,6 +216,34 @@ func (p *ReaderPool) NewBinaryReader(ctx context.Context, logger log.Logger, bkt
 	return reader, err
 }
 
+func (p *ReaderPool) tryEagerLoadIndexHeadersSnapshot(lazyBinaryReader *LazyBinaryReader, dir string, blockID ulid.ULID) {
```

Comment:
I think we need to handle failing to eagerly load index headers at a higher level. 

If I understand correctly, this will create and delete the marker file once per block, and that marker is scoped per-tenant. 

However, if store-gateway crashes while eagerly loading index headers (eg. because we OOMed), I'm thinking it's best to not eagerly load any index headers from any tenant when the store-gateway restarts. 

With the current setup, we'll only skip the tenants that were being loaded at the time of the crash, and still try to eagerly load all other tenants. However, imagine a scenario where the first tenant eagerly loads X GB, but the next tenant loads just enough to push the store-gateway over its memory limit, causing it to OOM and be killed. On the next run, we'll still load the first tenant's index headers, and only skip the second tenant.